src/main/java/org/mule/extension/file/internal/source/DirectoryListener.java

```diff
@@ -187,14 +191,88 @@ public void onTerminate() {}
 
   @Override
   public void poll(PollContext<InputStream, LocalFileAttributes> pollContext) {
-    PollWalker walker = new PollWalker(directoryPath, pollContext);
+
+    if (pollContext.isSourceStopping()) {
+      return;
+    }
+
+    LocalFileSystem fileSystem;
+    try {
+      fileSystem = fileSystemProvider.connect();
+    } catch (Exception e) {
+      LOGGER.error(format("Could not obtain connection while trying to poll directory '%s'. %s", directoryPath.toString(),
+                          e.getMessage()),
+                   e);
+      return;
+    }
+
     try {
-      walkFileTree(directoryPath, EnumSet.of(FOLLOW_LINKS), MAX_VALUE, walker);
+      List<Result<InputStream, LocalFileAttributes>> fileList =
+          fileSystem
+              .list(config, directoryPath.toString(), recursive, matcher,
+                    config.getTimeBetweenSizeCheckInMillis(timeBetweenSizeCheck, timeBetweenSizeCheckUnit).orElse(null));
+
+      if (fileList.isEmpty()) {
+        return;
+      }
+
+      for (Result<InputStream, LocalFileAttributes> file : fileList) {
+
+        LocalFileAttributes attributes = file.getAttributes().orElse(null);
+        if (attributes == null) {
+          continue;
+        }
+
+        if (attributes.isDirectory()) {
+          continue;
+        }
+
+        if (!matcher.test(attributes)) {
+          if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug("Skipping file '{}' because the matcher rejected it", attributes.getPath());
+          }
+          return;
+        }
+
+        if (!processFile(file, attributes, pollContext)) {
```

Comment:
is totally confusing, what one will think is that `processFile()` will return false if the process could not process the file correctly, but returns true or false depending if the source is stopping or not. 
Extract the boolean logic outside the `processFile()`